pkg/validators/am0010_k8s_resource_names.go

```diff
@@ -0,0 +1,234 @@
+
+	for _, m := range msgs {
+		msgStrings = append(msgStrings, string(m))
+	}
+
+	return failureMsg(strings.Join(msgStrings, ", "))
+}
+
+func Validatek8sResourceAndFieldNames(mb types.MetaBundle) types.ValidatorResult {
+	subValidators := []subValidator{
+		validateLabel,
+		validateTargetNamespace,
+		validateAllNamespaces,
+		validateCommonAnnotations,
+		validateCommonLabels,
+	}
+
+	var msgs []failureMsg
+
+	for _, v := range subValidators {
+		msgs = append(msgs, v(mb)...)
+	}
+
+	if len(msgs) > 0 {
+		return Fail(string(joinFailureMsgs(msgs...)))
+	}
+
+	return Success()
+}
+
+type subValidator func(types.MetaBundle) []failureMsg
+
+func validateLabel(mb types.MetaBundle) (failures []failureMsg) {
+	if msg := isValidLabelName(mb.AddonMeta.Label); !msg.IsEmpty() {
+		failures = append(failures, prefixedFailureMsg("label", msg))
+	}
+
+	return failures
+}
+
+func validateTargetNamespace(mb types.MetaBundle) (failures []failureMsg) {
+	if msg := isValidNamespaceName(mb.AddonMeta.TargetNamespace); !msg.IsEmpty() {
+		failures = append(failures, prefixedFailureMsg("targetNamespace", msg))
+	}
+
+	return failures
+}
+
+func validateAllNamespaces(mb types.MetaBundle) []failureMsg {
+	var result []failureMsg
+
+	if msgs := areValidNamespaceNames(mb.AddonMeta.Namespaces...); len(msgs) > 0 {
+		result = append(result, prefixedFailureMsg(
+			"namespaces", joinFailureMsgs(msgs...),
+		))
+	}
+
+	if msgs := areValidAnnotationNames(keys(mb.AddonMeta.NamespaceAnnotations)...); len(msgs) > 0 {
+		result = append(result, prefixedFailureMsg(
+			"namespaceAnnotations", joinFailureMsgs(msgs...),
+		))
+	}
+
+	if msgs := areValidLabelNames(keys(mb.AddonMeta.NamespaceLabels)...); len(msgs) > 0 {
+		result = append(result, prefixedFailureMsg(
+			"namespaceLabels", joinFailureMsgs(msgs...),
+		))
+	}
+
+	return result
+}
+
+func validateCommonAnnotations(mb types.MetaBundle) []failureMsg {
+	if mb.AddonMeta.CommonAnnotations == nil {
+		return []failureMsg{}
+	}
+
+	annotations := *mb.AddonMeta.CommonAnnotations
+
+	result := make([]failureMsg, 0, len(annotations))
+
+	if msgs := areValidAnnotationNames(keys(annotations)...); len(msgs) > 0 {
+		result = append(result, prefixedFailureMsg(
+			"commonAnnotations", joinFailureMsgs(msgs...),
+		))
+	}
+
+	return result
+}
+
+func validateCommonLabels(mb types.MetaBundle) []failureMsg {
+	if mb.AddonMeta.CommonLabels == nil {
+		return []failureMsg{}
+	}
+
+	labels := *mb.AddonMeta.CommonLabels
+
+	result := make([]failureMsg, 0, len(labels))
+
+	if msgs := areValidLabelNames(keys(labels)...); len(msgs) > 0 {
+		result = append(result, prefixedFailureMsg(
+			"commonLabels", joinFailureMsgs(msgs...),
+		))
+	}
+
+	return result
+}
+
+func areValidNamespaceNames(nsNames ...string) []failureMsg {
+	msgs := make([]failureMsg, 0, len(nsNames))
+
+	for _, name := range nsNames {
+		if msg := isValidNamespaceName(name); !msg.IsEmpty() {
+			msgs = append(msgs, msg)
+		}
+	}
+
+	return msgs
+}
```

Comment:
nit: refactor

`areValidNamespaceNames` + `areValidAnnotationNames` + `areValidLabelNames` have an identical implementation.

Maybe call a wrapper that takes a callback:

```go
func areValidNames(names ...string, callback func(string) failureMsg) []failureMsg {
    msgs := make([]failureMsg, 0, len(names))
    for _, name := rage names {
        if msg := callback(name); !msg.IsEmpty() {
            msgs := append(msgs, msg)
        }
    }
    return msgs
}

# then each `are...` funcs look like
func areValidAnnotationNames(aNames ...string) []failureMsg {
    return areValidNames(aNames, isValidAnnotationName)
}
```

Reply:
Hmm, yeah I guess I just stopped refactoring at some point lol. I'll give it a shot